### PR TITLE
consensus/docs: draft internal-CHANGELOG.md

### PR DESCRIPTION
**Pull request by nfrisby**
This PR adds a single new file, which contains its own motivation.

I had promised this file to @erikd a while ago, and @rvl just asked me for it. Rodney's request nicely motivated it by explaining that wallet, `db-sync`, etc may be integrating our changes especially long after we merge our PRs. So when I merge a PR and send a helpful Slack message warning them of breaking changes and how to integrate them, it's very easy for them to forget those notes event exist before they actually have to deal with the integration challenges.

### CODE REVIEW

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (rvl) [inline ouroboros-consensus/docs/internal-CHANGELOG.md:17]: Is it a problem to fix up or correct old changelog entries? It's always nice to get it right the first time, in the relevant PR, but I don't think anyone would mind previous entries being mutated.

I plan to read the changelog by running `git diff revA revB --  ouroboros-consensus/docs/internal-CHANGELOG.md`.

**Review by rvl — APPROVED** [review]
Thanks @nfrisby!

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (nc6) [reply ouroboros-consensus/docs/internal-CHANGELOG.md:17]: I guess the intention is to keep the alignment between `git blame` and the original commit?

**Review by nc6 — APPROVED** [review]
Think this is a great idea. It would be nice if we could include the PR number, but I appreciate that's particularly tricky!

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (dnadales) [inline ouroboros-consensus/docs/internal-CHANGELOG.md:28]: So in this case we won't be able to retrieve the commits related to the changes below based on this changelog blame. Is this an exception worth mentioning?

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (DavidEichmann) [inline ouroboros-consensus/docs/internal-CHANGELOG.md:51]: reason *to* reduce

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (DavidEichmann) [inline ouroboros-consensus/docs/internal-CHANGELOG.md:28]: It's unclear to me how we should use dates. This entry is large as it incorporates a bunch of recent changes. In the future are we intending to simple add a data and only 1 (or few) changes corresponding to e.g. a single PR? In that case we need need to worry too much about keeping everything immutable; we can just keep the date line immutable and look at the commit for the date line via git blame.

**Review by DavidEichmann — CHANGES_REQUESTED** [review]

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (DavidEichmann) [reply ouroboros-consensus/docs/internal-CHANGELOG.md:17]: @nc6 that's my understanding, but as i mention [below](https://github.com/input-output-hk/ouroboros-network/pull/3279#discussion_r679113587) I think we can weaken this requirement to simply the date lines being immutable.

**Review by Jasagredo — COMMENTED** [review]
I am okey with the idea of the ChangeLog although I would suggest a couple of improvements:
- Keep links to code/PRs as changes can span several PRs.
- Allow appending information to an entry to further refine it. If something is fundamentally wrong, probably just strikethrough it and add a reference to some explanation below, instead of deleting it.

**Comment by nfrisby** [comment]
Thanks for the comments everyone.

I think the immutability would be nice, but I likely overstated how important it is.

I'll update the draft based on the following notes, prompted by your comments. Thanks!

  - Downstream users deal with commit hashes. They have at least: the latest hash they've already integrated as well as the new hash they're considering how to integrate. This enables Rodney's clever plan of `git diff revA revB -- ouroboros-consensus/docs/internal-CHANGELOG.md`. I'll include that recommended command in the doc itself. If the user does that, then those of us writing the document can't really cause too much trouble unless we do something silly like change every line (eg DOS -> WIN line endings).
  
  - Moreover, downstream users deal with commit hashes from our mainline branch. That's where to find our "releases".
  
  - As of some point in our development history, all mainline commits are PR merge commits that list the PR number (ie we only use `bors` and never fast-forwards etc). Thus, indirectly, downstream users are picking some prefix of all of our merged PRs. But to emphasize: they don't ultimately care about PR numbers; they just want to know what changed between two hashes. The PR numbers would only help them (or someone else) do archaeology to figure out more details about a particular change. The hope is that the entries in this file will be complete enough that they won't have to do that.

  - I want the PR that makes the change to also add an entry to this log file. If we forget, and have to add the entry in a later PR, so be it -- we're human. But the goal is for the PR that makes the change to also update the log file. That would yield the most natural results.

  - We could put PR numbers in the doc itself: that number will already be known before the commit is finalized. But PR numbers are generally not of interest to our downstream users. And can be recovered more directly via a `git log` command.
  
    ```
    $ git log --first-parent --oneline 6cef814ca^..999f25035
    999f25035 Merge #3261
    f026e71e8 Merge #3278
    681351c06 Merge #3270
    36cb0b970 Merge #3230
    6cef814ca Merge #3275
    ```

  - Unlike PR numbers, we cannot predict the merge commit hash (that'd be self-referential). The idea to rely on `git blame` lets us pretend that the correct hash is automatically present in the file itself, since `git blame --first-parent` or `git blame` will display the relevant PR merge commit hash right next to each log entry.
  
  - I proposed including the date just so there is something in the file itself that indicates the time spans between entries. As David mentioned, it's not clear exactly which date this is/should be. Ideally, it's the timestamp of when the PR was merged; but that's self-referential. We could make the last step before merging the PR be to update the timestamp in the log, but that'll be easy to forget. So I think a practical and least-surprising/most-honest semantics is that the date in the file is "some time near" the PR merge timestamp.

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/internal-CHANGELOG.md]: I think the new text I just wrote does a better job of that than the old "bootstrapping" word I was hoping would cover it all by itself.

Please let me know if that's clearer now.

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/internal-CHANGELOG.md]: Thanks for the sharing the explicit command, Rodney. I found that very helpful.

Clarke: yes.

David: I'd say you're ultimately right about the dateline being the only necessarily immutable one, but I still think there is value in treating the entries as append-only.

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/internal-CHANGELOG.md]: I think my new introduction/overview section addresses this. Please let me know if not.

**Comment by nfrisby** [comment]
I've updated the PR, and I think the introduction/overview is much clearer now. Please let me know whether it addresses all your discussion points.

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (nfrisby) [inline ouroboros-consensus/docs/interface-CHANGELOG.md:1]: @Jasagredo said: 

> I am okey with the idea of the ChangeLog although I would suggest a couple of improvements:
>
> - Keep links to code/PRs as changes can span several PRs.
> - Allow appending information to an entry to further refine it. If something is fundamentally wrong, probably just strikethrough it and add a reference to some explanation below, instead of deleting it.

- Hmm. I think we should have one PR per log entry. We don't merge a PR until it builds, so at the very least some coherent portion of the ultimate change will already be in place. The log entry for each preparatory PR can at least list what changed and then defer more details by explicitly saying "a subsequent log entry will say more". How disruptive/challenging do you think that scheme would be for us?
- If we add the `<s>` and `</s>` tags on their own line, then we can strikethrough the surrounded lines without altering them, thereby preserving the `git blame` output 👍. I tested this a bit and if even works for surrounded headers as long as you add blanks lines around the strikethrough tags.

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (dnadales) [inline ouroboros-consensus/docs/interface-CHANGELOG.md:47]: at the top of the *old* log entry I assume. I wonder if this could be interpreted as adding new lines at the top of a new log entry.

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (dnadales) [inline ouroboros-consensus/docs/interface-CHANGELOG.md:50]:
```suggestion
oversimplifies the ...`. Such errata only corrects mistakes in the original log
```

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (dnadales) [reply ouroboros-consensus/docs/internal-CHANGELOG.md]: If you refer to this line:

> Now that this file exists,
>  entries added to it by later PRs will be narrower than this entry. This entry
>  should be the only one that is bootstrapping the file by discussing multiple
>  recent changes.

then yeah, it is clearer now. Thanks.

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/interface-CHANGELOG.md:50]: `errata` is plural, so `correct` is ... correct. Yeah?

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/interface-CHANGELOG.md]: Fixed, thanks. I changed `top of the log entry` to `top of that log entry`.

**Comment by nfrisby** [comment]
I think I've now responded to everyone's comments. Please have a look again if you recall unfinished business :)

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (dnadales) [reply ouroboros-consensus/docs/interface-CHANGELOG.md:50]: Oh, I see. Sorry about that.

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (DavidEichmann) [inline ouroboros-consensus/docs/interface-CHANGELOG.md:101]: I feel like part of this sentence was deleted. Do you mean "... useful to receive this flag in order to detect untrusted peers"?

Comment (DavidEichmann) [reply ouroboros-consensus/docs/interface-CHANGELOG.md:101]: ... or a missing "from"

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (DavidEichmann) [reply ouroboros-consensus/docs/internal-CHANGELOG.md]: @nfrisby absolutely agree

File: ouroboros-consensus/docs/internal-CHANGELOG.md
Comment (DavidEichmann) [reply ouroboros-consensus/docs/internal-CHANGELOG.md]: Perfect! Thanks :-)

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (Jasagredo) [inline ouroboros-consensus/docs/interface-CHANGELOG.md:51]:
```suggestion
entry; crucially, the code relevant to that entry is not changing. Such code
```

**Review by Jasagredo — APPROVED** [review]
Looks good to me

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/interface-CHANGELOG.md:101]: Fixed, thanks!

File: ouroboros-consensus/docs/interface-CHANGELOG.md
Comment (nfrisby) [reply ouroboros-consensus/docs/interface-CHANGELOG.md:51]: Fixed, thanks!

**Comment by nfrisby** [comment]
Hi @DavidEichmann. I think we're good to go here. Would you dismiss your Request Changes review? And then invoke `bors r+`? Thanks!

**Review by DavidEichmann — APPROVED** [review]

**Pull request merged**
